tests/checkpointing/test_model_checkpoint.py

```diff
@@ -938,3 +939,37 @@ def __init__(self, hparams):
     else:
         # make sure it's not AttributeDict
         assert type(ckpt[model.CHECKPOINT_HYPER_PARAMS_KEY]) == hparams_type
+
+
+def test_model_checkpoint_with_training_epoch_end(tmpdir):
+    """
+    This test ensures ModelCheckpoint issues a warning when the monitor is logged on training_epoch_end
+    """
+    class TestedModel(BoringModel):
+
+        def training_step(self, batch, batch_idx):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            self.log('train_loss', loss)
+            return {"loss": loss}
+
+        def training_epoch_end(self, outputs) -> None:
+            avg_loss = torch.stack([x["loss"] for x in outputs]).mean()
+            self.log('epoch_end_train_loss', avg_loss)
+
+    model = TestedModel()
+
+    chk = ModelCheckpoint(dirpath=tmpdir, monitor='epoch_end_train_loss', save_top_k=-1)
+    trainer = pl.Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=4,
+        progress_bar_refresh_rate=1,
+        callbacks=[chk],
+    )
+    trainer.current_epoch = 2
+    trainer.fit(model)
+
+    chks = os.listdir(tmpdir)
+    assert 'epoch=4.ckpt' not in chks
+    assert 'epoch=3.ckpt' not in chks
+    assert 'epoch=2.ckpt' not in chks
```

Comment:
This test is missing a `pytest.warns` check with the warning thrown :smile: 

Also it might be better to test

```os.listdir(tmpdir) == ['epoch=0.ckpt', ...]```

instead of testing that the rest don't exist

Reply:
The warning was removed. Also, I am thinking it might be better not to merge this PR as it is a wacky solution and work a on train_sanity check instead. @carmocca @rohitgr7 What are your thoughts?

Reply:
yes, let's close this one for now.